### PR TITLE
Add FP4 CuTe Flash Attention v4 backend for Blackwell

### PR DESCRIPTION
**Pull request by nholmber**
### Summary

Integrate the FP4 CUTE Flash Attention v4 kernel from [hao-ai-lab/flash-attention-fp4](https://github.com/hao-ai-lab/flash-attention-fp4) as a new attention backend (`FLASH_4_FP4`) for NVIDIA Blackwell GPUs (SM >= 10.0). This is the CUDA counterpart to the MXFP4 attention support on ROCm (`AITER_SAGE_V2`).

**Key design decisions:**
- **Runtime FP4 quantization via flashinfer**: Q and K are quantized to `Float4E2M1FN` at runtime using `flashinfer.quantization.nvfp4_quantize` with MMA-layout scale factors. V stays in BF16.
- **Why flashinfer over cutlass-dsl for quantization**: cutlass-dsl's `cute_tensor_like` + `convert_cute_tensor` also works and produces slightly faster kernel execution (~1-2%), but `convert_cute_tensor` calls `cute.testing.convert` -- a `@cute.jit` function that JIT-compiles a new CUDA kernel for every distinct tensor shape. During E2E inference with `torch.compile`, this triggers repeated JIT compilations (tens of seconds per step), making it unusable. flashinfer's `nvfp4_quantize` is a pre-compiled kernel with zero JIT overhead.
- **Important: cute-only install for the FP4 fork**: Only install `flash_attn/cute/` from the FP4 fork (`pip install -e .` in the cute subdirectory). Do NOT run the top-level `python setup.py install` or build the `hopper/` extension -- those compile CUDA extensions that cause SIGILL on Blackwell due to architecture mismatches with the fork's build configuration.
- **Hybrid attention schedule support**: Works with `AttentionSchedule` to use FP4 as the low-precision backend alongside cuDNN or FAv4 BF16 as the high-precision backend during early denoising steps.
- **`torch.compiler.disable`**: The FAv4 CUTE kernel uses `cutlass-dsl`'s `cute.compile()` to JIT-compile CUDA kernels at runtime, launching them via TVM FFI with raw pointer manipulation and a compile cache. None of this is traceable by `torch.compile`/Dynamo, so the attention call is wrapped with `@torch.compiler.disable`. This means the FP4 attention path runs outside the compiled graph -- the rest of the model (GEMMs, norms, etc.) is still compiled normally.

### Changes

| File | Description |
|------|-------------|
| `xfuser/core/distributed/attention_backend.py` | New `FLASH_4_FP4` enum value and registered `_flash_attn_4_fp4_call` handler. Conditional import of `flash_attn.cute.interface` and `fp4_quantize`. Handles permutation from xDiT's `(B,H,S,D)` to FAv4's `(B,S,H,D)` layout. |
| `xfuser/core/distributed/fp4_quantize.py` | New file. `quantize_qk_to_fp4()` uses `nvfp4_quantize` to produce packed `torch.float4_e2m1fn_x2` data + scale factors in the 7D MMA layout expected by the FAv4 kernel. Handles `nvfp4_quantize`'s internal padding of M to multiples of 128. |
| `xfuser/envs.py` | `_check_flash_attn_4_fp4()` detects FP4 kernel availability (import check + CUDA cap >= 10.0). Sets `CUTE_DSL_ENABLE_TVM_FFI=1` env var when FP4 is available. |
| `xfuser/core/distributed/runtime_state.py` | Compatibility checks: require Blackwell + FP4 kernel present. Block `ring_degree > 1` with FP4. Include FP4 in hybrid attention warnings. |

### Usage

```bash
# Pure FP4 attention
xdit --model Wan-AI/Wan2.2-I2V-A14B-Diffusers \
    --attention_backend flash_4_fp4 \
    --ulysses_degree 8 --use_torch_compile ...

# Hybrid: cuDNN for first N steps, FP4 for the rest
xdit --model Wan-AI/Wan2.2-I2V-A14B-Diffusers \
    --use_hybrid_attn_schedule \
    --hybrid_attn_high_precision_backend cudnn \
    --hybrid_attn_low_precision_backend flash_4_fp4 \
    --num_hybrid_attn_high_precision_steps 5 \
    --ulysses_degree 8 --use_torch_compile ...

# Combined: NVFP4 GEMMs + FP4 attention
xdit --model Wan-AI/Wan2.2-I2V-A14B-Diffusers \
    --use_fp4_gemms \
    --use_hybrid_attn_schedule \
    --hybrid_attn_high_precision_backend cudnn \
    --hybrid_attn_low_precision_backend flash_4_fp4 \
    --num_hybrid_attn_high_precision_steps 5 \
    --ulysses_degree 8 --use_torch_compile ...
```

### Dependencies

Requires the FP4 fork of flash-attention (only the CUTE subdirectory):
```bash
git clone https://github.com/hao-ai-lab/flash-attention-fp4.git
cd flash-attention-fp4/flash_attn/cute && pip install -e .
pip install "nvidia-cutlass-dsl==4.4.0.dev0" "apache-tvm-ffi==0.1.6" \
    "torch-c-dlpack-ext" "flashinfer-python"
```

### Microbenchmark Results (1xB200, Wan2.2 I2V shapes, seqlen=75600)

| Shape (B,S,H,D) | BF16 FAv4 | FP4 (flashinfer) | FP4 (cutlass-dsl) | Speedup (flashinfer) |
|------------------|-----------|-------------------|-------------------|---------------------|
| 1, 75600, 5, 128 (Ulysses=8) | 11.17 ms (1311 TF) | 8.65 ms (1691 TF) | 8.52 ms (1717 TF) | **1.29x** |
| 1, 75600, 40, 128 (Ulysses=1) | 91.90 ms (1274 TF) | 73.02 ms (1603 TF) | 72.08 ms (1624 TF) | **1.26x** |

> - `force_fp4_impl=True` (kernel-internal BF16->FP4 path) shows no speedup vs BF16, confirming that explicit quantization is essential.
> - cutlass-dsl quantization is ~1-2% faster at kernel level, but unusable for E2E inference due to `@cute.jit` JIT compilation overhead on every new tensor shape (tens of seconds per denoising step during warmup).
> - flashinfer's `nvfp4_quantize` is a pre-compiled kernel with zero JIT overhead -- the correct choice for production.

### E2E Validation Results (8xB200, torch.compile, Ulysses=8)

**Wan2.2 I2V 720x1280 81 frames, 40 steps:**

| Config | Time (s) | vs cuDNN BF16 |
|--------|----------|---------------|
| cuDNN BF16 (baseline) | 58.93 | 1.00x |
| FAv4 BF16 | 63.77 | 0.92x |
| FP4 pure | 55.14 | **1.07x** |
| Hybrid FP4 + cuDNN (5 hi-prec steps) | 55.92 | **1.05x** |
| Hybrid FP4 + FAv4 BF16 (5 hi-prec steps) | 57.01 | 1.03x |

**Wan2.2 T2V 720x1280 81 frames, 40 steps:**

| Config | Time (s) | vs cuDNN BF16 |
|--------|----------|---------------|
| cuDNN BF16 (baseline) | 57.86 | 1.00x |
| FAv4 BF16 | 62.66 | 0.92x |
| FP4 pure | 55.14 | **1.05x** |
| Hybrid FP4 + cuDNN (5 hi-prec steps) | 55.92 | **1.03x** |
| Hybrid FP4 + FAv4 BF16 (5 hi-prec steps) | 57.01 | 1.01x |

**Flux2 t2i 1024x1024, 50 steps:**

| Config | Time (s) | vs cuDNN BF16 |
|--------|----------|---------------|
| cuDNN BF16 (baseline) | 2.14 | 1.00x |
| FAv4 BF16 | 2.56 | 0.84x |
| FP4 pure | 2.72 | 0.79x |
| Hybrid FP4 + cuDNN (5 hi-prec steps) | 2.58 | 0.83x |
| Hybrid FP4 + FAv4 BF16 (5 hi-prec steps) | 2.50 | 0.86x |

**Key observations:**
 - FP4 attention provides **1.05-1.07x speedup** on Wan2.2 (both I2V and T2V) due to large sequence lengths (75600 tokens per GPU with Ulysses=8).
- Flux2 sequences are too short for FP4 to overcome the quantization + kernel launch overhead.

### Quality

bf16 Wan2.2 
https://github.com/user-attachments/assets/33cf3a11-df93-4c59-9d58-2497944c3769

fp4 Wan2.2 (hybrid 5 steps)
https://github.com/user-attachments/assets/429f10a3-f425-4454-9331-046e3b85e383

### Test plan

- [x] Microbenchmark: FP4 vs BF16 on Wan2.2 shapes (nheads=5 and nheads=40)
- [x] E2E: Wan2.2 I2V 5 configs (cuDNN BF16, FAv4 BF16, FP4 pure, hybrid cuDNN, hybrid FAv4)
- [x] E2E: Wan2.2 T2V 5 configs
- [x] E2E: Flux2 t2i 5 configs

### CODE REVIEW

**Review by avjves — COMMENTED** [review]
Looks good!

The BS > 1 bug Gemini mentioned seems to be real though. I think we should either fix it or add a check for BS size that will fail accordingly.

File: xfuser/core/distributed/fp4_quantize.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    sf = (
+        sf_data
+        .reshape(total_m, total_k, 32, 4, 4)
+        .reshape(batch, rest_m, nheads, rest_k, 32, 4, 4)
+        .permute(0, 2, 1, 3, 4, 5, 6)
+        .contiguous()
+        .permute(4, 5, 2, 6, 3, 1, 0)
+    )
+
+    return fp4, sf
```
[the remark this answers is not in the document]
Comment (avjves) [reply xfuser/core/distributed/fp4_quantize.py:85]: So the fix for BS>1 would be:
```
  batch, seqlen, nheads, headdim = tensor.shape

  # Pad seqlen so each batch element aligns independently
  padded_seqlen = math.ceil(seqlen / 128) * 128
  if padded_seqlen > seqlen:
      tensor = torch.nn.functional.pad(tensor, (0, 0, 0, 0, 0, padded_seqlen - seqlen))

  global_sf = torch.ones(1, device=tensor.device, dtype=torch.float32)
  fp4_data, sf_data = nvfp4_quantize(
      tensor.reshape(batch * padded_seqlen, nheads * headdim),
      global_sf,
      sfLayout=SfLayout.layout_128x4,
      do_shuffle=False,
  )

  fp4 = (
      fp4_data
      .reshape(batch, padded_seqlen, nheads, headdim // 2)
      .view(torch.int8)
      .view(torch.float4_e2m1fn_x2)
  )
  # Slice back to original seqlen for the kernel
  if padded_seqlen > seqlen:
      fp4 = fp4[:, :seqlen, ...].contiguous()

  # sf uses padded dimensions - rest_m is now exact division
  rest_m = padded_seqlen // 128
  sf_kph = headdim // sf_vec_size
  rest_k = sf_kph // 4
  total_m = batch * rest_m
  total_k = (nheads * sf_kph) // 4

  sf = (
      sf_data
      .reshape(total_m, total_k, 32, 4, 4)
      .reshape(batch, rest_m, nheads, rest_k, 32, 4, 4)
      .permute(0, 2, 1, 3, 4, 5, 6)
      .contiguous()
      .permute(4, 5, 2, 6, 3, 1, 0)
  )

  return fp4, sf
```

I don't have access to blackwell to test it, though. @nholmber do you have access to test?

[the remark this answers is not in the document]
Comment (nholmber) [reply xfuser/core/distributed/fp4_quantize.py:85]: Thanks! Committed the change after validating with BS=2 and seqlen that's not divisible by 128

**Review by avjves — APPROVED** [review]
